fix: Correctness in `get_role_permissions` and `has_perm` JS APIs

marination opened the pull request:
Solves part 1 of https://github.com/frappe/frappe/pull/17621

## **Issue:**
#### Inconsistency
- `get_perm` uses `get_role_permissions` to get perm-level wise rights
- `get_role_permissions` return value is expected to have `if_owner` as a property so that `get_perm` can add or remove permissions depending on document ownership. 
- `if_owner` is not considered in `get_role_permissions` > so the return value does not have `if_owner` > so `get_perm` never proceeds with giving rights depending on ownership

> This return value inconsistent with `get_role_permissions()` in **permissions.py**

#### Stale Values
- `has_perm` could be invoked with or without a document object
-  In both cases it checked `frappe.perm.doctype_perm[doctype]` which is clearly a **DocType level** window property
-  If this pre-generated property did not exist, it calls `get_perm` and sets the result in the namespace
-  This result could be DocType level perms or Document level perms (these will have `if_owner` and user perms applied too). Either way it is set at the **DocType level** window property
- Any consecutive `has_perm` calls that occur without a reload will get stale values even if they are seeking document level perms

> There are no issues so far as `has_perm` is mostly invoked for DocType level permissions (`doc` isn't passed), but this behaviour is wrong and can be confusing if someone tries using `has_perm` for Document level permissions. `form.js` cleverly generates fresh document level permissions via `get_perm` so this doesn't impact it


## **Fix:**
- Account for `if_owner` as a property if it is enabled in the doctype role permission
- This is used in `get_perm` if the document is passed to it to give document ownership priveledges
- Regenerate doc level perms in `has_perm`

marination committed:
fix: Consistency in `get_role_permissions` return value

- Return value contains `if_owner` propert in object same as py
- Elaborate code documentation

marination committed:
fix: Don't assign document level perms to doctype level permission store

- If `doc` is passed to `has_perm`, checking for pre-stored doctype level perms is wrong
- It gives back stale values that don't consider document level permissions and only change on reload (window property)
- Get freshly evaluate perms at the doc level
- If no `doc` is involved, doctype level custom window property can be used

sagarvora commented on frappe/public/js/frappe/model/perm.js:
```diff
@@ -117,20 +119,38 @@ $.extend(frappe.perm, {
 					let value = perm[permlevel][right] || p[right] || 0;
-					if (value) {
+
+					if (p.if_owner && value) {
+						// if_owner is enabled for perm,
+						// construct perm object inside "if_owner" property
+						if (!perm[permlevel]["if_owner"]) {
+							perm[permlevel]["if_owner"] = {}
+						}
+						perm[permlevel]["if_owner"][right] = value;
```
This won't work because `if_owner` should only be applied if all permissions for this right have `if_owner` checked. In other words, if any permission for this right doesn't have `if_owner` checked, `if_owner` shouldn't apply.

Also, `if_owner` is only required for permlevel 0.

Fixing in new commit.

sagarvora committed:
fix: correct logic for `if_owner` and refactor

sagarvora commented:
## Summary of changes in my commit

### Fixes
- created a set for rights without `if_owner` instead of creating an `if_owner` object. consequently, fixed existing usage of `if_owner` object.
### Refactor
- `perm[0]` => `base_perm`
- simpler `has_perm`
- loop for repetitive code and flatter code for share permissions in `get_perm`

sagarvora commented:
I think this logic is correct now, but we should probably add some tests ❔

sagarvora commented:
other thoughts (optional):
- we should probably cache the output of `get_role_permissions` instead of  / in addition to the current `doctype_perm`, because a majority of calls have `doc` specified
- for administrator, all rights should be allowed instead of current "read" only behavior in `get_perm` (probably another PR)

marination committed:
test: Cypress test to check basic `has_perm` and `get_perm` JS APIs

- Test removes System Manager role for user and expects certain perms
- On Kanban Board DocType, only System Manager is allowed to print/export/email/report/share
- DocType level test only

marination committed:
Merge branch 'develop' into get_role_permissions-js-consistency

marination committed:
chore: Cache Document level perms as well in `has_perm`

- Cache documet level output of `get_perm` in `has_perm` as well

marination committed:
Merge branch 'develop' into get_role_permissions-js-consistency

marination commented:
@sagarvora 
> other thoughts (optional):
> 
> * we should probably cache the output of `get_role_permissions` instead of  / in addition to the current `doctype_perm`, because a majority of calls have `doc` specified

Done in this PR

> * for administrator, all rights should be allowed instead of current "read" only behavior in `get_perm` (probably another PR)

Created https://github.com/frappe/frappe/issues/18958

marination committed:
chore: `web_form.py` format via pre-commit

marination committed:
fix: Avoid caching unsaved documents and secure whitelist decorator

- Avoid caching documents like 'new-item-1', default to doctype perms instead
- Use secure `whitelist_for_tests` decorator instead

sagarvora committed:
chore: cache `role_perms` instead

sagarvora commented:
## Why am I caching role perms instead?
- Similar as earlier behaviour of caching `doctype_perm`
- Cache is available even when `doc` is passed, or when `get_perm` is called directly

sagarvora committed:
fix: decorator usage

sagarvora committed:
chore: move cheaper condition first

sagarvora committed:
chore: revert to `doctype_perm` cache for now
